llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,359 @@
+    Type *T = (*OpIt)->getType();
+    Aspects = GetAspectsFromType(T, Types, Cache);
+    Result.insert(Aspects.begin(), Aspects.end());
+  }
+
+  return Result;
+}
+
+class MissedAspectDiagnosticInfo : public DiagnosticInfo {
+  Twine Msg;
+
+public:
+  MissedAspectDiagnosticInfo(Twine DiagMsg,
+                             DiagnosticSeverity Severity = DS_Warning)
+      : DiagnosticInfo(DK_Linker, Severity), Msg(std::move(DiagMsg)) {}
+
+  void print(DiagnosticPrinter &DP) const override { DP << Msg; }
+};
+
+void emitWarning(LLVMContext &C, const StringRef Msg) {
+  C.diagnose(MissedAspectDiagnosticInfo(Msg));
+}
+
+template <class Container> std::string Join(const Container &C, char sep) {
+  std::string S;
+  auto it = C.begin();
+  for (size_t i = 0; i != C.size(); ++i, ++it) {
+    if (i > 0)
+      S += ", ";
+
+    S += std::to_string(*it);
+  }
+
+  return S;
+}
+
+template <class Container>
+void CheckDeclaredAspects(LLVMContext &C, const Function *F,
+                          const Container &Aspects) {
+  MDNode *MDN = F->getMetadata("intel_declared_aspects");
+  if (!MDN)
+    return;
+
+  AspectsTy DeclaredAspects;
+  for (unsigned i = 0; i != MDN->getNumOperands(); ++i) {
+    const ConstantAsMetadata *CM =
+        dyn_cast<const ConstantAsMetadata>(MDN->getOperand(i));
+    DeclaredAspects.insert(
+        dyn_cast<const ConstantInt>(CM->getValue())->getSExtValue());
+  }
+
+  SmallVector<int> MissedAspects;
+  for (int aspect : Aspects) {
+    if (DeclaredAspects.count(aspect) == 0) {
+      MissedAspects.push_back(aspect);
+    }
+  }
+
+  if (!MissedAspects.empty()) {
+    std::string AspectsStr = Join(MissedAspects, ',');
+    // TODO: demangle function name and aspect's IDs?
+    std::string Msg = formatv(
+        "for function \"{0}\" there is the list of missed aspects: [{1}]",
+        F->getName(), AspectsStr);
+
+    emitWarning(C, Msg);
+  }
+}
+
+using FunctionToAspectsMapTy = DenseMap<Function *, SmallSet<int, 4>>;
+using CallGraphTy = DenseMap<Function *, SmallPtrSet<Function *, 8>>;
+
+void CreateAspectMetadataForFunctions(FunctionToAspectsMapTy &Map) {
+  for (auto &it : Map) {
+    Function *F = it.first;
+    auto &Aspects = it.second;
+    if (Aspects.empty())
+      continue;
+
+    LLVMContext &C = F->getContext();
+    std::vector<Metadata *> AspectsMetadata;
+    AspectsMetadata.reserve(Aspects.size());
+    for (int aspect : Aspects) {
+      AspectsMetadata.push_back(ConstantAsMetadata::get(
+          ConstantInt::getSigned(Type::getInt32Ty(C), aspect)));
+    }
+
+    MDNode *MDN = MDNode::get(C, AspectsMetadata);
+    F->setMetadata("intel_used_aspects", MDN);
+  }
+}
+
+void CheckAllDeclaredAspects(FunctionToAspectsMapTy &Map) {
+  for (auto &it : Map) {
+    Function *F = it.first;
+    auto &Aspects = it.second;
+    if (Aspects.empty())
+      continue;
+
+    LLVMContext &C = F->getContext();
+    CheckDeclaredAspects(C, F, Aspects);
```

Comment:
```suggestion
    CheckUsedAndDeclaredAspects(C, F, Aspects);
```

I think that we primarily checking _used_ aspects, because declaration of an aspect alone can't trigger any diagnostics, but a _use_ of undeclared aspect is what triggers the diagnostic message